papers/hatra23/hatra23.tex

```diff
@@ -0,0 +1,107 @@
+Roblox~\cite{Roblox} platform for shared 3D experiences.  Luau extends
+the Lua~\cite{Lua} language, notably by providing type-driven tooling
+such as autocomplete and API documentation (as well as traditional type
+error reporting). Roblox has hundreds of millions of users, and
+millions of creators, ranging from children learning to program for
+the first time to professional development studios.
+
+In HATRA 2021, we presented \emph{The Goals Of The Luau Type
+System}~\cite{BFJ21:GoalsLuau}, describing the human factors issues
+with designing a type system for a language with a heterogeneous
+developer community. The design flows from the needs of the different
+communities: beginners are focused on immediate goals (``the stairs
+should light up when a player walks on them'') and less on the code
+quality concerns of more experienced developers; for all users
+type-driven tooling is important for productivity. These needs result in a design with two modes:
+\begin{itemize}
+\item \emph{non-strict mode}, aimed at non-professionals, focused on
+  minimizing false positives, and
+\item \emph{strict mode}, aimed at professionals, focused on
+  minimizing false negatives (i.e.~type soundness).
+\end{itemize}
+For both communities, type-driven tooling is important, so we
+provide \emph{infallible type inference}, which infers types no matter 
+the inputs.
+
+\section{New stuff}
+
+The main improvement is that Luau now has support for \emph{semantic
+subtyping}~\cite{Jef22:SemanticSubtyping}. Semantic subtyping
+interprets types as sets of values, and subtyping as set
+inclusion~\cite{GF05:GentleIntroduction}. This is aligned with the
+\emph{minimize false positives} goal of Luau non-strict mode, since
+semantic subtyping only reports a failure of subtyping when there is a
+value which inhabits the candidate subtype, but not the candidate
+supertype. This has the added benefit of improving error reporting in
+the prototype implementation: since the prototype is constructive, the
+witness for the failure of subtyping can help drive error reports.
+
+The other big improvement is how we view gradual typing, Rather than
+the approach of~\cite{ST07:GradualTyping}, which uses \emph{consistent
+subtyping} where $\ANY \lesssim T \lesssim \ANY$ for any type $T$, we
+adopt an approach based on \emph{error suppression}, where $\ANY$ is
+an error-suppressing type, and any failures of subtyping involving
+error-suppressing types are not reported. Users can explicitly
+suppress type errors by declaring variables with type $\ANY$, and
+since an expression with a type error has an error-suppressing type we
+avoid cascading errors. Error-suppression is in production Luau, and is
+mechanically verified~\cite{BJ23:agda-typeck}.
+
+\section{Further work}
+
+Currently the type inference system uses greedy inference, which is
+very sensitive to order of declarations, and can easily result in
+false positives. This needs replaced by some form of local type
```

Comment:
'needs to be'?

Reply:
You can take the boy out of Glasgow but you can't take Glasgow out of the boy.

Reply:
TIL.